File: fibonacci.cpp
#include <iostream>
#include <string>
#include <random>

struct FibonacciCounter {
    int count0 = 0;
    int count1 = 0;
    
    int fibonacci(int n) {
        if (n == 0) {
            count0++;
            return 0;
        } else if (n == 1) {
            count1++;
            return 1;
        } else {
            return fibonacci(n - 1) + fibonacci(n - 2);
        }
    }
};

int getRandomNumber() {
    std::random_device rd; 
    std::mt19937 mt(rd()); 
    std::uniform_int_distribution<int> dist(0, 40); 

    return dist(mt); 
}

int getNumIterationsFromUser() {
    int numIterations;
    while (true) {
        std::cout << "Enter the number of iterations: ";
        std::cin >> numIterations;

        if (std::cin.fail()) {
            std::cin.clear(); 
            std::cin.ignore(std::numeric_limits<std::streamsize>::max(), '\n'); 
            std::cout << "Invalid input. Please enter a valid integer." << std::endl;
        }        
        else if (numIterations <= 0) {
            std::cout << "Please enter a positive number of iterations." << std::endl;
        }
        else {
            break; 
        }
    }
    return numIterations;
}

FibonacciCounter getFibonacciCounts(int n) {
    FibonacciCounter counter;
    counter.fibonacci(n);
    return counter;
}

void printFibonacciCounts(int i) {
    int randomNum = getRandomNumber();
    FibonacciCounter result = getFibonacciCounts(randomNum);
    // 과제 출력 요청되로 출력되는 횟수를 공백으로 구분하여서 출력하였습니다. 좀더 구체적인 정보를 입력한 방법도 구현해 보았습니다(line 62)
    std::cout << result.count0 << " " << result.count1 << std::endl;
    // std::cout << "Result for round " << i << " = " << randomNum << ": " << result.count0 << " " << result.count1 << std::endl;
}

int main() {
    int numIterations = getNumIterationsFromUser();
    for (int i = 1; i <= numIterations; ++i) {
       printFibonacciCounts(i);
    }
    return 0;
}

File: fibonacci.js
const readline = require('readline')

class FibonacciCounter {
  constructor() {
    this.count0 = 0
    this.count1 = 0
  }

  fibonacci(n) {
    if (n === 0) {
      this.count0++
      return 0
    } else if (n === 1) {
      this.count1++
      return 1
    } else {
      return this.fibonacci(n - 1) + this.fibonacci(n - 2)
    }
  }
}

function generateRandomNum() {
  // min, max value could be converted into parameters to make the function generic.
  const min = 0
  const max = 40
  return Math.floor(Math.random() * (max - min) + min)
}

// Get count how many times 0 and 1 are called.
function getFibonacciCounts(n) {
  const counter = new FibonacciCounter()
  counter.fibonacci(n)
  return { count0: counter.count0, count1: counter.count1 }
}

// 'i' is not utilized as it solely displays counts.
function printFibonacciCounts(i) {
  const randomNum = generateRandomNum()
  const result = getFibonacciCounts(randomNum)

  // 과제 출력 요청되로 출력되는 횟수를 공백으로 구분하여서 출력하였습니다. 좀더 구체적인 정보를 입력한 방법도 구현해 보았습니다(line 44-46)
  console.log(`${result.count0} ${result.count1}`)

  //   console.log(
  //     `Result for round ${i} = ${randomNum} (count0 : ${result.count0} & count1 : ${result.count1})`
  //   )
}

const rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout,
})

rl.question('Enter the number of iterations: ', (numIterations) => {
  rl.close()

  numIterations = parseInt(numIterations)

  if (isNaN(numIterations) || numIterations <= 0) {
    console.error('Please enter a valid positive number of iterations.')
    return
  }

  for (let i = 1; i <= numIterations; i++) {
    printFibonacciCounts(i)
  }
})
